### PR TITLE
feat(java client): Add connector functions to java client

### PR DESCRIPTION
**Pull request by jzaralim**
### Description 
Adds `createConnector`, `dropConnector`, `listConnector` and `describeConnector` to the Java client. I can split this into separate PRs if that makes it easier to review.

### Testing done 
Unit and integration tests

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/ConnectorDescription.java
Comment (vcrfxia) [inline ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/ConnectorDescription.java:28]: What sources are these? External sources (as opposed to ksqlDB streams/tables)?

File: ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/ConnectorDescription.java
Comment (vcrfxia) [inline ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/ConnectorDescription.java:34]: "consumed" meaning this is only present for sink connectors?

File: ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/ConnectorDescription.java
Comment (vcrfxia) [inline ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/ConnectorDescription.java:20]: This interface should contain the connector name as well. I think we should simply have it extend `ConnectorInfo` and add the two additional fields (sources and topics).

File: ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/ConnectorType.java
Comment (vcrfxia) [inline ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/ConnectorType.java:19]: We should add a javadoc about what situations might cause the connector type to be `UNKNOWN`, since I imagine will be confused about this otherwise.

File: ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/Client.java
Comment (vcrfxia) [inline ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/Client.java:248]: @colinhicks do you want to sanity-check these new Java client interfaces for managing connectors, from the user/product perspective?

File: ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/impl/ConnectorCommandResponseHandler.java
Comment (vcrfxia) [inline ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/impl/ConnectorCommandResponseHandler.java:107]: A bit late for this PR but we've been wanting to change the pattern of how the Java client response interfaces relate to the server response objects for a while now, in order to avoid custom JSON parsing like this. Here's the ticket with context: https://github.com/confluentinc/ksql/issues/6042

We should really do this before the next time we add additional interfaces, but it's outside the scope for this PR. (If you're keen on helping clean up the codebase and want to tackle it as a follow-up PR that'd be super, but no pressure 😁 )

File: ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/impl/ConnectorDescriptionImpl.java
Comment (vcrfxia) [inline ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/impl/ConnectorDescriptionImpl.java:39]: nit: add second argument to name failure should this ever happen (and similarly for below and other classes).
```suggestion
    this.connectorClass = Objects.requireNonNull(connectorClass, "connectorClass");
```

File: ksqldb-api-client/src/test/java/io/confluent/ksql/api/client/integration/ClientIntegrationTest.java
Comment (vcrfxia) [inline ksqldb-api-client/src/test/java/io/confluent/ksql/api/client/integration/ClientIntegrationTest.java:1034]: If all three of these tests create the same connector, why do whichever tests run second and third not fail (because the connector already exists)?

File: ksqldb-api-client/src/test/java/io/confluent/ksql/api/client/integration/ClientIntegrationTest.java
Comment (vcrfxia) [inline ksqldb-api-client/src/test/java/io/confluent/ksql/api/client/integration/ClientIntegrationTest.java:1067]: Add validation that the connector was successfully created into the "Given" here, else the test could be a no-op (as far as the reader is aware).

File: ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/Client.java
Comment (vcrfxia) [inline ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/Client.java:273]: nit: extra space before period.

**Review by vcrfxia — COMMENTED** [review]
Hey @jzaralim -- awesome test coverage! Are you planning to add docs in a separate PR? If you wouldn't mind, we should also add docs for the new `serverInfo()` method you added as part of the 0.16 release (could be a separate PR).

Left some questions inline, mostly about the interfaces, but the high-level LGTM!

File: ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/CreateConnectorResult.java
Comment (jzaralim) [reply ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/CreateConnectorResult.java:20]: I'm not sure, but this is the response that Connect returns when creating a new connector. If this is not useful for the Java client, we can remove it.

File: ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/ConnectorDescription.java
Comment (jzaralim) [reply ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/ConnectorDescription.java]: Updated the Javadoc to make this more clear.

File: ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/ConnectorDescription.java
Comment (jzaralim) [reply ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/ConnectorDescription.java]: Updated the Javadoc to make this more clear.

File: ksqldb-api-client/src/test/java/io/confluent/ksql/api/client/integration/ClientIntegrationTest.java
Comment (jzaralim) [reply ksqldb-api-client/src/test/java/io/confluent/ksql/api/client/integration/ClientIntegrationTest.java]: The statement fails but doesn't throw an error. I ended up moving this to setUp and tearDown though.

File: ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/CreateConnectorResult.java
Comment (jzaralim) [reply ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/CreateConnectorResult.java:20]: Hm actually I guess this wouldn't be the only place the client shows less info than the old rest api. I'll remove it, having 3 very similar classes for connector responses is clunky anyway.

File: ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/ConnectorType.java
Comment (vcrfxia) [inline ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/ConnectorType.java:23]: Is this the only reason the connector type might be `UNKNOWN` or are there others as well, for example, when listing existing connectors? For some reason I thought there were other possible reasons for an `UNKNOWN` type as well but I could be wrong.

File: ksqldb-api-client/src/test/java/io/confluent/ksql/api/client/integration/ClientIntegrationTest.java
Comment (vcrfxia) [inline ksqldb-api-client/src/test/java/io/confluent/ksql/api/client/integration/ClientIntegrationTest.java:286]: I don't think this belongs here. Most tests don't need the connector so it feels wasteful. How about we wrap this in a `givenConnectorExists()` method and only call it from the `Given:` portion of the tests that actually need it instead?

File: ksqldb-api-client/src/test/java/io/confluent/ksql/api/client/integration/ClientIntegrationTest.java
Comment (vcrfxia) [inline ksqldb-api-client/src/test/java/io/confluent/ksql/api/client/integration/ClientIntegrationTest.java:304]: How does this work if it's being called after `REST_APP.getServiceContext().close();`? 

Instead of trying to put this in `tearDown()`, we can wrap it into `givenConnectorExists()` (see comment above) and `classTearDown()` instead.

**Review by vcrfxia — APPROVED** [review]
Thanks @jzaralim -- LGTM once the inline comments are addressed and docs are added. It'd be nice to get another opinion on the new client interfaces as well cc @colinhicks .

File: ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/ConnectorType.java
Comment (jzaralim) [reply ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/ConnectorType.java]: I couldn't find any other references to `UNKNOWN`, but there could be a reason on the Connect side of things that I'm missing. Updated docs to be more accurate

File: ksqldb-api-client/src/test/java/io/confluent/ksql/api/client/integration/ClientIntegrationTest.java
Comment (vcrfxia) [inline ksqldb-api-client/src/test/java/io/confluent/ksql/api/client/integration/ClientIntegrationTest.java:1078]: Is this exception expected (under normal cases)? What would cause this?

File: ksqldb-api-client/src/test/java/io/confluent/ksql/api/client/integration/ClientIntegrationTest.java
Comment (vcrfxia) [inline ksqldb-api-client/src/test/java/io/confluent/ksql/api/client/integration/ClientIntegrationTest.java:1139]: Same questions as above. I notice there's no analogous try-catch around the `SHOW CONNECTORS;` request in the `cleanupConnectors()` helper method, curious why this situation is different.

**Review by vcrfxia — COMMENTED** [review]
Two more minor questions inline -- LGTM overall! (Plus the comments from before on docs and getting product input on the new interfaces) Thanks @jzaralim .

File: ksqldb-api-client/src/test/java/io/confluent/ksql/api/client/integration/ClientIntegrationTest.java
Comment (jzaralim) [reply ksqldb-api-client/src/test/java/io/confluent/ksql/api/client/integration/ClientIntegrationTest.java:1139]: It takes a while to provision a connector, even after `CREATE CONNECTOR` returns a successful response. When ksqlDB calls `SHOW CONNECTORS;`, it makes a call to '/info' for reach connector, so if the new connector is not done provisioning, it returns an error response which propagates.

File: ksqldb-api-client/src/test/java/io/confluent/ksql/api/client/integration/ClientIntegrationTest.java
Comment (jzaralim) [reply ksqldb-api-client/src/test/java/io/confluent/ksql/api/client/integration/ClientIntegrationTest.java:1078]: It's not. It's just that since this is a part of a lambda, we need to do a try-catch here in order for the code to compile.

File: ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/Client.java
Comment (derekjn) [inline ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/Client.java:249]: I believe we use `Map<String, Object>` to pass properties elsewhere in the client interface. Is there a reason for using `<String, String>` here, or should we keep this consistent with the other interface methods?

Comment (jzaralim) [reply ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/Client.java:249]: Good catch! Will update the PR.

**Pull request merged**
